proj2/src/wavcb.cpp

```diff
@@ -36,83 +36,63 @@ int main(int argc, char *argv[]) {
     }
     
     // initialize centroids
-    int index, cur_centroidvec_size;
     vector<vector<short>> centroids;
+    {set<int> indexesUsed;
+    int index, prev_centroidvec_size;
     for(int i=0; i<numCentroids; i++) {
-        cur_centroidvec_size = centroids.size();
-        while(centroids.size() == cur_centroidvec_size) {
+        prev_centroidvec_size = centroids.size();
+        while(centroids.size() == prev_centroidvec_size) {
             index = rand() % blocks.size();
-            if(!count(centroids.begin(),centroids.end(),blocks.at(index))) { // if centroids does not yet contain that possible centroid
-                centroids.push_back(blocks.at(index));
+            if (indexesUsed.count(index) == 0) {
+                centroids.emplace_back(blocks.at(index));
+                indexesUsed.insert(index);
             }
         }
-    }
-    /*cout << blocks.size() << endl;
-    for(int i=0; i<numCentroids; i++) { cout << to_string(centroids.at(i).at(0)) << endl; }*/
-
-
-    /*
-        Código daqui para baixo não testado, penso que tem erros no formato dos blocks e dos centroids
-    */
-
-    // classify blocks (apply k-means algorithm)
-    vector<vector<short>> closest_centroids;
-    vector<long> smallest_errors;
-    long error;
-    long smallest_local_error;
-    vector<short> local_centroid;
-    for(int i=0; i<blocks.size(); i++) {
-        smallest_local_error = calcEn(blocks.at(i), centroids.at(0));
-        local_centroid = centroids.at(0);
-        for(int j=1; j<centroids.size(); j++) {
-            error = calcEn(blocks.at(i), centroids.at(j));
+    }}
+
+    // generates centroids (apply k-means algorithm)
+    {vector<vector<vector<short>*>> closest_blocks(centroids.size());
```

Comment:
To know which centroid is closer to the block in the x-th position, you set the x-th postion of the `closest_centroids` equal to the closest centroid `vector`.
I took an inverse approach. Each centroid (mapped to each postion of the 1st layer vector) has several blocks that he is closest to (second layer vector stores a pointer to the block's `vector`)